core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -0,0 +1,132 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.avro;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import java.util.Deque;
+import java.util.List;
+import org.apache.avro.Schema;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+public abstract class AvroSchemaWithTypeVisitor<T> {
+  public static <T> T visit(org.apache.iceberg.Schema iSchema, Schema schema, AvroSchemaWithTypeVisitor<T> visitor) {
+    return visit(iSchema.asStruct(), schema, visitor);
+  }
+
+  public static <T> T visit(Type iType, Schema schema, AvroSchemaWithTypeVisitor<T> visitor) {
+    switch (schema.getType()) {
+      case RECORD:
+        return visitRecord(iType != null ? iType.asStructType() : null, schema, visitor);
+
+      case UNION:
+        return visitUnion(iType, schema, visitor);
+
+      case ARRAY:
+        return visitArray(iType, schema, visitor);
+
+      case MAP:
+        Types.MapType map = iType != null ? iType.asMapType() : null;
+        return visitor.map(map, schema,
+            visit(map != null ? map.valueType() : null, schema.getValueType(), visitor));
+
+      default:
+        return visitor.primitive(iType != null ? iType.asPrimitiveType() : null, schema);
+    }
+  }
+
+  private static <T> T visitRecord(Types.StructType struct, Schema record, AvroSchemaWithTypeVisitor<T> visitor) {
+    // check to make sure this hasn't been visited before
+    String name = record.getFullName();
+    Preconditions.checkState(!visitor.recordLevels.contains(name),
+        "Cannot process recursive Avro record %s", name);
+
+    visitor.recordLevels.push(name);
+
+    List<Schema.Field> fields = record.getFields();
+    List<String> names = Lists.newArrayListWithExpectedSize(fields.size());
+    List<T> results = Lists.newArrayListWithExpectedSize(fields.size());
+    for (Schema.Field field : fields) {
+      int fieldId = AvroSchemaUtil.getFieldId(field);
+      Types.NestedField iField = struct != null ? struct.field(fieldId) : null;
+      names.add(field.name());
+      results.add(visit(iField != null ? iField.type() : null, field.schema(), visitor));
+    }
+
+    visitor.recordLevels.pop();
+
+    return visitor.record(struct, record, names, results);
+  }
+
+  private static <T> T visitUnion(Type type, Schema union, AvroSchemaWithTypeVisitor<T> visitor) {
+    List<Schema> types = union.getTypes();
+    List<T> options = Lists.newArrayListWithExpectedSize(types.size());
+    for (Schema branch : types) {
+      if (branch.getType() == Schema.Type.NULL) {
+        options.add(visit((Type) null, branch, visitor));
+      } else {
+        options.add(visit(type, branch, visitor));
+      }
+    }
+    return visitor.union(type, union, options);
+  }
+
+  private static <T> T visitArray(Type type, Schema array, AvroSchemaWithTypeVisitor<T> visitor) {
+    if (array.getLogicalType() instanceof LogicalMap) {
```

Comment:
Is it worth while to also check for `AvroSchemaUtil.isKeyValueSchema(array.getElementType())`?

Reply:
No, but I think it would be good to check whether the Iceberg type is a map.